### PR TITLE
SONAR-8980 Add organization to user ES index

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java [hunk 1]
```diff
@@ -55,4 +58,21 @@ public void deleteByOrganizationUuid(DbSession dbSession, String organizationMem
     return mapper(dbSession).selectOrganizationUuidsByUser(userId);
   }
 
+  /**
+   *
+   * @param loginOrganizationConsumer {@link BiConsumer}<String,String> (login, organization uuid)
+   */
+  public void selectForUserIndexing(DbSession dbSession, List<String> logins, BiConsumer<String, String> loginOrganizationConsumer) {
+    executeLargeInputsWithoutOutput(logins, list -> mapper(dbSession).selectForIndexing(list)
+      .forEach(row -> loginOrganizationConsumer.accept(row.get("login"), row.get("organizationUuid"))));
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: Interesting usage of BiConsumer ! I wasn't aware of this class, but it's good to know,

Comment (julienlancelot) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: I'm just a little worried about the mix usage of user id and user login in this DAO, but I understand than using user id in the 2 new classes would add complexity into UserResultSetIterator.
In fact we should replace user_id by login in tables, but this is out of scope of this PR.

File: server/sonar-db-dao/src/test/java/org/sonar/db/organization/OrganizationMemberDaoTest.java [hunk 2]
```diff
@@ -88,6 +92,38 @@ public void select_organization_uuids_by_user_id() {
...
   @Test
+  public void select_for_indexing() {
+    OrganizationDto org1 = db.organizations().insert(o -> o.setUuid("ORG_1"));
+    OrganizationDto org2 = db.organizations().insert(o -> o.setUuid("ORG_2"));
+    UserDto user1 = db.users().insertUser("L_1");
+    UserDto user2 = db.users().insertUser("L_2");
+    db.organizations().addMember(org1, user1);
+    db.organizations().addMember(org1, user2);
+    db.organizations().addMember(org2, user1);
+    List<Tuple> result = new ArrayList<>();
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/test/java/org/sonar/db/organization/OrganizationMemberDaoTest.java]: I like this way of writing unit tests :)

File: server/sonar-server/src/test/java/org/sonar/server/user/index/UserIndexTest.java [hunk 3]
```diff
@@ -144,36 +143,27 @@ public void searchUsers() throws Exception {
...
-    assertThat(underTest.search(userQuery.setLogins(singletonList("USER_")).build(), new SearchOptions()).getDocs()).isEmpty();
-    assertThat(underTest.search(userQuery.setLogins(emptyList()).build(), new SearchOptions()).getDocs()).isEmpty();
+  public void search_users_filter_by_organization_uuid() {
+    esTester.putDocuments(INDEX_TYPE_USER.getIndex(), INDEX_TYPE_USER.getType(), newUser(USER1_LOGIN, asList("user_1", "u1")).setEmail("email1")
+      .setOrganizationUuids(newArrayList("O1", "O2")));
+    esTester.putDocuments(INDEX_TYPE_USER.getIndex(), INDEX_TYPE_USER.getType(), newUser(USER2_LOGIN, emptyList()).setEmail("email2")
+      .setOrganizationUuids(newArrayList("O2")));
+
+    assertThat(underTest.search(userQuery.setOrganizationUuid("O42").build(), new SearchOptions()).getDocs()).isEmpty();
+    assertThat(underTest.search(userQuery.setOrganizationUuid("O2").build(), new SearchOptions()).getDocs()).hasSize(2);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/user/index/UserIndexTest.java]: It would make tests clearer to replace the check on size of the list by checking elements in the list : 
assertThat(underTest.search(userQuery.setOrganizationUuid("O2").build(), new SearchOptions()).getDocs()).extracting(UserLogin).containsOnly(USER1_LOGIN, USER2_LOGIN);

File: server/sonar-server/src/test/resources/org/sonar/server/user/index/UserResultSetIteratorTest/shared.xml [hunk 4]
```diff
@@ -1,35 +0,0 @@
-<dataset>
```
Comment (julienlancelot) [inline server/sonar-server/src/test/resources/org/sonar/server/user/index/UserResultSetIteratorTest/shared.xml]: Great !!!

File: server/sonar-server/src/test/java/org/sonar/server/organization/ws/RemoveMemberActionTest.java [hunk 5]
```diff
@@ -324,12 +338,14 @@ private TestResponse call(@Nullable String organizationKey, @Nullable String log
     return request.execute();
   }
 
-  private void assertNotAMember(String organizationUuid, int userId) {
-    assertThat(dbClient.organizationMemberDao().select(dbSession, organizationUuid, userId)).isNotPresent();
+  private void assertNotAMember(String organizationUuid, UserDto user) {
+    assertThat(dbClient.organizationMemberDao().select(dbSession, organizationUuid, user.getId())).isNotPresent();
+    assertThat(userIndex.getNullableByLogin(user.getLogin()).organizationUuids()).doesNotContain(organizationUuid);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/organization/ws/RemoveMemberActionTest.java]: As userIndex.getNullableByLogin will be removed someday, it would be best to use es search, for instance : 
    assertThat(
      esTester.client()
        .prepareSearch(INDEX_TYPE_USER)
        .setQuery(termQuery(FIELD_ORGANIZATIONS, nameQuery))
        .get()
        .getHits()
        .getTotalHits()).isEqualTo(0);

File: server/sonar-server/src/main/java/org/sonar/server/user/index/UserResultSetIterator.java [hunk 6]
```diff
@@ -79,15 +81,29 @@ static UserResultSetIterator create(DbClient dbClient, DbSession session, @Nulla
...
 
-  private static void setParameter(PreparedStatement stmt, @Nullable String login) throws SQLException {
-    if (login != null) {
-      stmt.setString(1, login);
+  private static void setParameters(PreparedStatement stmt, @Nullable List<String> logins) throws SQLException {
+    if (logins == null) {
+      return;
+    }
+
+    for (int i = 0; i < logins.size(); i++) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/user/index/UserResultSetIterator.java]: No stream API ?

File: server/sonar-server/src/test/java/org/sonar/server/user/index/UserIndexerTest.java [hunk 7]
```diff
@@ -67,18 +69,39 @@ public void index_everything() {
...
-  private UserIndexer createIndexer() {
-    return new UserIndexer(db.getDbClient(), es.client());
+  @Test
+  public void index_single_user() {
+    UserDto user = db.users().insertUser();
+    UserDto anotherUser = db.users().insertUser();
+
+    underTest.index(user.getLogin());
+
+    List<UserDoc> docs = es.getDocuments(UserIndexDefinition.INDEX_TYPE_USER, UserDoc.class);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/user/index/UserIndexerTest.java]: As this method will probably disappear someday, it would be best to replace it by an es search, for instance : 
```
    assertThat(
      esTester.client()
        .prepareSearch(INDEX_TYPE_USER)
        .setQuery(termQuery(FIELD_ORGANIZATIONS, nameQuery))
        .get()
        .getHits()
        .getTotalHits()).isEqualTo(1);
```

File: server/sonar-server/src/test/java/org/sonar/server/organization/ws/DeleteActionTest.java [hunk 8]
```diff
@@ -293,6 +306,8 @@ public void request_also_deletes_members_of_specified_organization() {
     assertThat(db.getDbClient().organizationMemberDao().select(db.getSession(), org.getUuid(), user1.getId())).isNotPresent();
     assertThat(db.getDbClient().organizationMemberDao().select(db.getSession(), org.getUuid(), user2.getId())).isNotPresent();
     assertThat(db.getDbClient().organizationMemberDao().select(db.getSession(), otherOrg.getUuid(), user1.getId())).isPresent();
+    assertThat(userIndex.getNullableByLogin(user1.getLogin()).organizationUuids()).doesNotContain(org.getUuid()).contains(otherOrg.getUuid());
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/organization/ws/DeleteActionTest.java]: Same remark as in previous commit, replace this method usage

File: server/sonar-server/src/test/java/org/sonar/server/organization/OrganizationCreationImplTest.java [hunk 9]
```diff
@@ -213,11 +230,17 @@ public void create_creates_default_template_for_new_organization() throws Organi
...
+
     mockForSuccessfulInsert(SOME_UUID, SOME_DATE);
+    userIndexer.index(someUser.getLogin());
 
-    underTest.create(dbSession, SOME_USER_ID, FULL_POPULATED_NEW_ORGANIZATION);
+    underTest.create(dbSession, someUser, FULL_POPULATED_NEW_ORGANIZATION);
 
-    assertThat(dbClient.organizationMemberDao().select(dbSession, SOME_UUID, SOME_USER_ID)).isPresent();
+    assertThat(dbClient.organizationMemberDao().select(dbSession, SOME_UUID, someUser.getId())).isPresent();
+    assertThat(userIndex.getNullableByLogin(someUser.getLogin()).organizationUuids()).contains(SOME_UUID);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/organization/OrganizationCreationImplTest.java]: Please replace userIndex.getNullableByLogin

File: server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java [hunk 10]
```diff
@@ -55,4 +58,21 @@ public void deleteByOrganizationUuid(DbSession dbSession, String organizationMem
     return mapper(dbSession).selectOrganizationUuidsByUser(userId);
   }
 
+  /**
+   *
+   * @param loginOrganizationConsumer {@link BiConsumer}<String,String> (login, organization uuid)
```
Comment (teryk) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: I want to specify the first and second element of the BiConsumer (login, organization uuid). Otherwise, it's not obvious

File: server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java (hunk 1 above)
Comment (teryk) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: should we replace user_id by login? We've just created ORGANIZATION_MEMBERS and we put the user_id

File: server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java [hunk 11]
```diff
@@ -55,4 +58,21 @@ public void deleteByOrganizationUuid(DbSession dbSession, String organizationMem
...
+   */
+  public void selectForUserIndexing(DbSession dbSession, List<String> logins, BiConsumer<String, String> loginOrganizationConsumer) {
+    executeLargeInputsWithoutOutput(logins, list -> mapper(dbSession).selectForIndexing(list)
+      .forEach(row -> loginOrganizationConsumer.accept(row.get("login"), row.get("organizationUuid"))));
+  }
+
+  /**
+   *
+   * @param loginOrganizationConsumer {@link BiConsumer}<String,String> (login, organization uuid)
```
Comment (teryk) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: same, I want to document that the BiConsumer has login, then organization uuid

File: server/sonar-server/src/main/java/org/sonar/server/user/index/UserResultSetIterator.java (hunk 6 above)
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/user/index/UserResultSetIterator.java]: nope because `PreparedStatement.setString(…)` can throw a checked exception and you cannot have it simply in a lambda

File: server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java (hunk 10 above)
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: Ok, I thought it was automatically generated, I haven't paid attention to the end

File: server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java (hunk 1 above)
Comment (julienlancelot) [inline server/sonar-db-dao/src/main/java/org/sonar/db/organization/OrganizationMemberDao.java]: Hum, indeed, but Simon told us not to do that.
